# Bike rental data Cleaning 

### Library and Data Imports

In [172]:
import pandas as pd
import numpy as np
import glob

In [237]:
#creating list of filenames using glob library
files = glob.glob("bike-rental-starter-kit/data/JC-******-citibike-tripdata.csv")
#iterating through the list of filenames and appending data frames to the list
df_list=[]
for filename in files:
    data = pd.read_csv(filename)
    df_list.append(data)
bikes_df = pd.concat(df_list)
#importing the weather file
weather_df = pd.read_csv('bike-rental-starter-kit/data/newark_airport_2016.csv')


### Citi Bike Data Overview

In [207]:
bikes_df.head(3)

Trip Duration           Start Time            Stop Time  Start Station ID  \
0            361  2016-02-01 00:31:18  2016-02-01 00:37:19              3202   
1            297  2016-02-01 01:55:05  2016-02-01 02:00:02              3195   
2           1155  2016-02-01 02:40:05  2016-02-01 02:59:20              3183   

  Start Station Name  Start Station Latitude  Start Station Longitude  \
0       Newport PATH               40.727224               -74.033759   
1            Sip Ave               40.730743               -74.063784   
2     Exchange Place               40.716247               -74.033459   

   End Station ID End Station Name  End Station Latitude  \
0            3203    Hamilton Park             40.727596   
1            3194  McGinley Square             40.725340   
2            3210   Pershing Field             40.742677   

   End Station Longitude  Bike ID   User Type  Birth Year  Gender  
0             -74.044247    24393  Subscriber      1975.0       1  
1             -74.067622    24394  Subscriber      1985.0       2  
2             -74.051789    24676  Subscriber      1976.0       1

In [234]:
bikes_df.tail(3)

trip_duration          start_time           stop_time  \
247569           2048 2016-05-31 23:25:28 2016-05-31 23:59:36   
247570            455 2016-05-31 23:31:57 2016-05-31 23:39:32   
247571            239 2016-05-31 23:47:38 2016-05-31 23:51:37   

        start_station_id start_station_name  start_station_latitude  \
247569              3199       Newport Pkwy               40.728745   
247570              3220  5 Corners Library               40.734961   
247571              3185          City Hall               40.717732   

        start_station_longitude  end_station_id end_station_name  \
247569               -74.032108            3199     Newport Pkwy   
247570               -74.059503            3215      Central Ave   
247571               -74.043845            3211       Newark Ave   

        end_station_latitude  end_station_longitude  bike_id   user_type  \
247569             40.728745             -74.032108    24573  Subscriber   
247570             40.746730             -74.049251    24561  Subscriber   
247571             40.721525             -74.046305    24685  Subscriber   

        birth_year  gender  trip_minutes  trip_hours  valid_duration   age  \
247569      1975.0       1         34.13        0.57            True  48.0   
247570      1964.0       1          7.58        0.13            True  59.0   
247571      1993.0       1          3.98        0.07            True  30.0   

            id  
247569  247569  
247570  247570  
247571  247571

* There are 19487 rows in the data set
* Trip Duration is in seconds
* Gender can be 0=unknown, 1=male, and 2=female
* User Type has values Customer=24 hour pass or 3 day user and Subscriber=Annual Member

In [209]:
bikes_df.describe()

Trip Duration  Start Station ID  Start Station Latitude  \
count   2.475840e+05     247584.000000           247584.000000   
mean    8.856305e+02       3207.065206               40.723121   
std     3.593798e+04         26.955103                0.008199   
min     6.100000e+01       3183.000000               40.692640   
25%     2.480000e+02       3186.000000               40.717732   
50%     3.900000e+02       3201.000000               40.721525   
75%     6.660000e+02       3211.000000               40.727596   
max     1.632981e+07       3426.000000               40.752559   

       Start Station Longitude  End Station ID  End Station Latitude  \
count            247584.000000   247584.000000         247584.000000   
mean                -74.046438     3203.572553             40.722594   
std                   0.011211       61.579494              0.007958   
min                 -74.096937      147.000000             40.692216   
25%                 -74.050656     3186.000000             40.716540   
50%                 -74.044247     3199.000000             40.721124   
75%                 -74.038051     3211.000000             40.727224   
max                 -74.032108     3426.000000             40.801343   

       End Station Longitude        Bike ID     Birth Year         Gender  
count          247584.000000  247584.000000  228585.000000  247584.000000  
mean              -74.045855   24935.260481    1979.335276       1.123534  
std                 0.011283     748.469712       9.596809       0.518687  
min               -74.096937   14552.000000    1900.000000       0.000000  
25%               -74.050444   24491.000000    1974.000000       1.000000  
50%               -74.043117   24609.000000    1981.000000       1.000000  
75%               -74.036486   24719.000000    1986.000000       1.000000  
max               -73.957390   27274.000000    2000.000000       2.000000

In [210]:
print("Trip Duration 5 max:")
display(bikes_df['Trip Duration'].sort_values(ascending=False).head())
print("Birth Dates 5 min:")
display(bikes_df['Birth Year'].sort_values().head())

Trip Duration 5 max:


8903     16329808
18510     4826890
3233      2104123
3269      2100551
3371      2071209
Name: Trip Duration, dtype: int64

Birth Dates 5 min:


4417     1900.0
22758    1934.0
32771    1937.0
6163     1937.0
2929     1937.0
Name: Birth Year, dtype: float64

* max Trip Duration looks like outlier
* min Birth Year is very small

In [211]:
# checking for missing values
bikes_df.isna().sum()

Trip Duration                  0
Start Time                     0
Stop Time                      0
Start Station ID               0
Start Station Name             0
Start Station Latitude         0
Start Station Longitude        0
End Station ID                 0
End Station Name               0
End Station Latitude           0
End Station Longitude          0
Bike ID                        0
User Type                    380
Birth Year                 18999
Gender                         0
dtype: int64

* User type - mising data
* Birth year - missing data
* Gender - data are not missing, but we know from data dictionary, that O represent "unknown"

In [212]:
# checking for duplicates
bikes_df.duplicated().sum()

0

We have no duplicates, which is good

In [213]:
bikes_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 247584 entries, 0 to 19487
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Trip Duration            247584 non-null  int64  
 1   Start Time               247584 non-null  object 
 2   Stop Time                247584 non-null  object 
 3   Start Station ID         247584 non-null  int64  
 4   Start Station Name       247584 non-null  object 
 5   Start Station Latitude   247584 non-null  float64
 6   Start Station Longitude  247584 non-null  float64
 7   End Station ID           247584 non-null  int64  
 8   End Station Name         247584 non-null  object 
 9   End Station Latitude     247584 non-null  float64
 10  End Station Longitude    247584 non-null  float64
 11  Bike ID                  247584 non-null  int64  
 12  User Type                247204 non-null  object 
 13  Birth Year               228585 non-null  float64
 14  Gende

Almost all datatypes are good, except Start and Stop time

### Citi Bike Data Transformation

#### Columns to investigate:

* Start Time/Stop Time: wrong data types.
* Trip Duration: looks like there are outliers.
* Birth Year: missing data and possible outliers.
* Gender: investigate unknowns.
* User type: missing data.


#### Changing columns names for comfortable work

In [214]:
bikes_df.columns = bikes_df.columns.str.lower().str.replace(' ', '_')
bikes_df.columns

Index(['trip_duration', 'start_time', 'stop_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'birth_year', 'gender'],
      dtype='object')

Looks good, let's start transformation

#### Start Time/ Stop Time

In [215]:
# casting objects as timestamps
bikes_df.start_time = pd.to_datetime(bikes_df.start_time,infer_datetime_format=True)
bikes_df.stop_time = pd.to_datetime(bikes_df.stop_time,infer_datetime_format=True)
bikes_df.dtypes

trip_duration                       int64
start_time                 datetime64[ns]
stop_time                  datetime64[ns]
start_station_id                    int64
start_station_name                 object
start_station_latitude            float64
start_station_longitude           float64
end_station_id                      int64
end_station_name                   object
end_station_latitude              float64
end_station_longitude             float64
bike_id                             int64
user_type                          object
birth_year                        float64
gender                              int64
dtype: object

#### Trip Duration

In [216]:
# creating an easier to interpret trip time column in minutes and hours
bikes_df['trip_minutes'] = round(bikes_df['trip_duration'] / 60, 2)
bikes_df['trip_hours'] = round(bikes_df['trip_duration']/(60**2),2)
bikes_df[['trip_duration','trip_minutes','trip_hours']].describe()

trip_duration   trip_minutes     trip_hours
count   2.475840e+05  247584.000000  247584.000000
mean    8.856305e+02      14.760503       0.246006
std     3.593798e+04     598.966283       9.982777
min     6.100000e+01       1.020000       0.020000
25%     2.480000e+02       4.130000       0.070000
50%     3.900000e+02       6.500000       0.110000
75%     6.660000e+02      11.100000       0.180000
max     1.632981e+07  272163.470000    4536.060000

* minimum trip length = 61 seconds
* maximum trip = 4,536 hours =189 days

According to the data dictionary, trips under 60 seconds were already assumed to be false starts/stops and were pruned.

For the max, Citi Bike policies state that the maximum length of a trip is 24 hours (or 86400 seconds).

We won't delete the longer trips, since it is possible these correspond to bike docking issues or users breaking the rules (which analysts might want to investigate). However, we will add a flag so that the analytics team can easily filter these out if desired.

In [217]:
# create a column to flag rides that do not comply with system protocols
bikes_df['valid_duration'] = bikes_df.trip_duration.apply(lambda x: 0 if x > 86400 else 1).astype(bool)

#### Birth Year

In [218]:
#calculating age from birth years
bikes_df['age'] = pd.Timestamp.now().year - bikes_df.birth_year
bikes_df.age.sort_values(ascending=False).head(10)

4417     123.0
22758     89.0
6163      86.0
2929      86.0
32771     86.0
16672     86.0
23319     83.0
2201      83.0
29718     83.0
7314      82.0
Name: age, dtype: float64

smb is 123 years old in 2023 year, so I guess we can delete this row from the observation.

In [219]:
bikes_df.drop(4417, inplace = True)

Now let's have a look at missing data

In [220]:
birth_year_missing_data = bikes_df[bikes_df.birth_year.isnull()]

In [221]:
birth_year_missing_data.head()

trip_duration          start_time           stop_time  start_station_id  \
125           3665 2016-02-01 09:56:46 2016-02-01 10:57:52              3212   
148           1081 2016-02-01 11:43:51 2016-02-01 12:01:53              3183   
154           1100 2016-02-01 12:04:35 2016-02-01 12:22:56              3192   
163           1579 2016-02-01 12:26:09 2016-02-01 12:52:29              3192   
168           1619 2016-02-01 12:53:44 2016-02-01 13:20:44              3186   

     start_station_name  start_station_latitude  start_station_longitude  \
125     Christ Hospital               40.734786               -74.050444   
148      Exchange Place               40.716247               -74.033459   
154  Liberty Light Rail               40.711242               -74.055701   
163  Liberty Light Rail               40.711242               -74.055701   
168       Grove St PATH               40.719586               -74.043117   

     end_station_id    end_station_name  end_station_latitude  \
125            3185           City Hall             40.717732   
148            3192  Liberty Light Rail             40.711242   
154            3192  Liberty Light Rail             40.711242   
163            3183      Exchange Place             40.716247   
168            3203       Hamilton Park             40.727596   

     end_station_longitude  bike_id user_type  birth_year  gender  \
125             -74.043845    24531  Customer         NaN       0   
148             -74.055701    24624  Customer         NaN       0   
154             -74.055701    24624  Customer         NaN       0   
163             -74.033459    24624  Customer         NaN       0   
168             -74.044247    24510  Customer         NaN       0   

     trip_minutes  trip_hours  valid_duration  age  
125         61.08        1.02            True  NaN  
148         18.02        0.30            True  NaN  
154         18.33        0.31            True  NaN  
163         26.32        0.44            True  NaN  
168         26.98        0.45            True  NaN

As I can see from these 5 rows: In every row, where birth_year is missing, gender is missing also. Additionally, user_type for this rows is always 'Customer'. Let's check if this pattern will cover the rest of the dataset.

In [222]:
print('Length of the Dataframe with missing birth years:')
display(len(birth_year_missing_data))
print('User type value counts:')
display(birth_year_missing_data.user_type.value_counts(dropna=False))
print('Gender value counts:')
display(birth_year_missing_data.gender.value_counts())

Length of the Dataframe with missing birth years:


18999

User type value counts:


Customer      15470
Subscriber     3529
Name: user_type, dtype: int64

Gender value counts:


0    18999
Name: gender, dtype: int64

* As we can see: all genders are missing.That means that if there is no birth_year ther is also no gender data.
* the customer user_type seems to be missing a lot of birth_year data
Let's check the percentages for user_type:

In [223]:
round(birth_year_missing_data['user_type'].value_counts() / bikes_df['user_type'].value_counts(),4)*100

Customer      99.67
Subscriber     1.52
Name: user_type, dtype: float64

* 99% of customers are missing birth_year
* 1.5% of subscribers are missing birth_year
* all records with missing birth_year also have unknown gender, which likely also corresponds to the customer user_type
We won't delete records that are missing birth_year, since that would remove all customers. We'll have to keep this missing data in mind when developing the database.

#### user_type

In [224]:
round(bikes_df.user_type.value_counts(dropna=False) / len(bikes_df.user_type),4)*100

Subscriber    93.58
Customer       6.27
NaN            0.15
Name: user_type, dtype: float64

Less then 1% is missing, so actually we can delete these rows, but let's have a deeper look, maybe birth_year is also missing and we will be able to consider missing user_types as 'Customers'.

In [225]:
missing_user_type_data = bikes_df[bikes_df.user_type.isnull()]

In [226]:
missing_user_type_data.birth_year.isnull().sum()

0

There is no missing birth_year, so missing user_types are not 'Customers'. We can delete this rows, but I will better mark them as unknown:

In [227]:
bikes_df.fillna({'user_type':'Unknown'},inplace=True)

In [233]:
bikes_df.reset_index(drop=True,inplace=True)
bikes_df['id'] = bikes_df.index
bikes_df.tail(3)

trip_duration          start_time           stop_time  \
247569           2048 2016-05-31 23:25:28 2016-05-31 23:59:36   
247570            455 2016-05-31 23:31:57 2016-05-31 23:39:32   
247571            239 2016-05-31 23:47:38 2016-05-31 23:51:37   

        start_station_id start_station_name  start_station_latitude  \
247569              3199       Newport Pkwy               40.728745   
247570              3220  5 Corners Library               40.734961   
247571              3185          City Hall               40.717732   

        start_station_longitude  end_station_id end_station_name  \
247569               -74.032108            3199     Newport Pkwy   
247570               -74.059503            3215      Central Ave   
247571               -74.043845            3211       Newark Ave   

        end_station_latitude  end_station_longitude  bike_id   user_type  \
247569             40.728745             -74.032108    24573  Subscriber   
247570             40.746730             -74.049251    24561  Subscriber   
247571             40.721525             -74.046305    24685  Subscriber   

        birth_year  gender  trip_minutes  trip_hours  valid_duration   age  \
247569      1975.0       1         34.13        0.57            True  48.0   
247570      1964.0       1          7.58        0.13            True  59.0   
247571      1993.0       1          3.98        0.07            True  30.0   

            id  
247569  247569  
247570  247570  
247571  247571

### Weather Data Overview

In [229]:
weather_df.head(3)

STATION                                         NAME        DATE  \
0  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-01   
1  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-02   
2  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-03   

    AWND  PGTM  PRCP  SNOW  SNWD  TAVG  TMAX  TMIN  TSUN  WDF2   WDF5  WSF2  \
0  12.75   NaN   0.0   0.0   0.0    41    43    34   NaN   270  280.0  25.9   
1   9.40   NaN   0.0   0.0   0.0    36    42    30   NaN   260  260.0  21.0   
2  10.29   NaN   0.0   0.0   0.0    37    47    28   NaN   270  250.0  23.9   

   WSF5  
0  35.1  
1  25.1  
2  30.0

In [231]:
weather_df.tail(3)

STATION                                         NAME        DATE  \
363  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-12-29   
364  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-12-30   
365  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-12-31   

      AWND  PGTM  PRCP  SNOW  SNWD  TAVG  TMAX  TMIN  TSUN  WDF2   WDF5  WSF2  \
363   8.05   NaN  0.36   0.0   0.0    38    45    31   NaN   170  150.0  18.1   
364  14.99   NaN  0.00   0.0   0.0    37    42    32   NaN   270  270.0  25.9   
365  12.30   NaN  0.00   0.0   0.0    35    44    29   NaN   200  220.0  21.9   

     WSF5  
363  25.1  
364  33.1  
365  28.0

In [232]:
weather_df.describe()

AWND  PGTM        PRCP        SNOW        SNWD        TAVG  \
count  366.000000   0.0  366.000000  366.000000  366.000000  366.000000   
mean     9.429973   NaN    0.104945    0.098087    0.342623   57.196721   
std      3.748174   NaN    0.307496    1.276498    2.078510   17.466981   
min      2.460000   NaN    0.000000    0.000000    0.000000    8.000000   
25%      6.765000   NaN    0.000000    0.000000    0.000000   43.000000   
50%      8.720000   NaN    0.000000    0.000000    0.000000   56.000000   
75%     11.410000   NaN    0.030000    0.000000    0.000000   74.000000   
max     22.820000   NaN    2.790000   24.000000   20.100000   89.000000   

             TMAX        TMIN  TSUN        WDF2        WDF5        WSF2  \
count  366.000000  366.000000   0.0  366.000000  364.000000  366.000000   
mean    65.991803   48.459016   NaN  217.841530  228.269231   20.484426   
std     18.606301   17.135790   NaN  102.548282   97.415777    6.848390   
min     18.000000    0.000000   NaN   10.000000   10.000000    6.900000   
25%     51.250000   35.000000   NaN  150.000000  150.000000   15.000000   
50%     66.000000   47.000000   NaN  240.000000  260.000000   19.900000   
75%     83.000000   64.000000   NaN  300.000000  300.000000   23.900000   
max     99.000000   80.000000   NaN  360.000000  360.000000   48.100000   

             WSF5  
count  364.000000  
mean    26.801648  
std      8.882610  
min     10.100000  
25%     19.900000  
50%     25.100000  
75%     31.100000  
max     66.000000

### Cleaning and Transforming Weather Data

PGTM (peak gust time) and TSUN (total minutes of sun) are entirely missing and will be dropped. WDFx and WSFx refer to the fastest x-minute wind, and are difficult to connect directly to our data (since we don't know when that speed occurred.) We'll therefore stick with AWND (average wind speed) for our wind-speed variable.

Also, since all our data is from the same station, we'll drop those details for now.

In [238]:
weather_df.drop(['STATION','NAME','PGTM','TSUN', 'WDF2', 'WDF5', 'WSF2', 'WSF5'],axis=1, inplace=True)

Let's rename the columns to make the data clearer

In [241]:
weather_df.columns = ['rec_date','avg_wind','prcp','snow_amt','snow_depth','tavg','tmax','tmin']

Lastly, let's add some new columns to assist analysts

In [243]:
# create convenience binary columns for weather events
weather_df['rain'] = weather_df['prcp'].apply(lambda x: 1 if x > 0 else 0).astype(bool)
weather_df['snow'] = weather_df['snow_amt'].apply(lambda x: 1 if x > 0 else 0).astype(bool)

In [244]:
weather_df.head()

rec_date  avg_wind  prcp  snow_amt  snow_depth  tavg  tmax  tmin   rain  \
0  2016-01-01     12.75   0.0       0.0         0.0    41    43    34  False   
1  2016-01-02      9.40   0.0       0.0         0.0    36    42    30  False   
2  2016-01-03     10.29   0.0       0.0         0.0    37    47    28  False   
3  2016-01-04     17.22   0.0       0.0         0.0    32    35    14  False   
4  2016-01-05      9.84   0.0       0.0         0.0    19    31    10  False   

    snow  
0  False  
1  False  
2  False  
3  False  
4  False